# Pandas Data Analysis Part 1
**Friday, 25 November 2022**

In [1]:
import pandas as pd 
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

In [2]:
type(df)

pandas.core.frame.DataFrame

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [5]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [6]:
# initial insite
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Age and Cabin less number record so there is null value in it.

In [7]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [8]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [9]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [10]:
df.dtypes == "object"

PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [11]:
type(df.dtypes)

pandas.core.series.Series

series similar to list

In [12]:
df.dtypes[df.dtypes == "object"]

Name        object
Sex         object
Ticket      object
Cabin       object
Embarked    object
dtype: object

In [13]:
col_name = df.dtypes[df.dtypes == "object"].index
col_name

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [14]:
df[col_name]

Name     Sex  \
0                              Braund, Mr. Owen Harris    male   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                               Heikkinen, Miss. Laina  female   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                             Allen, Mr. William Henry    male   
..                                                 ...     ...   
886                              Montvila, Rev. Juozas    male   
887                       Graham, Miss. Margaret Edith  female   
888           Johnston, Miss. Catherine Helen "Carrie"  female   
889                              Behr, Mr. Karl Howell    male   
890                                Dooley, Mr. Patrick    male   

               Ticket Cabin Embarked  
0           A/5 21171   NaN        S  
1            PC 17599   C85        C  
2    STON/O2. 3101282   NaN        S  
3              113803  C123        S  
4              373450   NaN        S  
..                ...   ...      ...  
886            211536   NaN        S  
887            112053   B42        S  
888        W./C. 6607   NaN        S  
889            111369  C148        C  
890            370376   NaN        Q  

[891 rows x 5 columns]

In [15]:
df[col_name].describe()

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

In [16]:
# pandas profiling is more accurate 

In [17]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# add a new column 


In [18]:
df['ineuron']= 'Sharad'
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked ineuron  
0        0         A/5 21171   7.2500   NaN        S  Sharad  
1        0          PC 17599  71.2833   C85        C  Sharad  
2        0  STON/O2. 3101282   7.9250   NaN        S  Sharad  
3        0            113803  53.1000  C123        S  Sharad  
4        0            373450   8.0500   NaN        S  Sharad  
..     ...               ...      ...   ...      ...     ...  
886      0            211536  13.0000   NaN        S  Sharad  
887      0            112053  30.0000   B42        S  Sharad  
888      2        W./C. 6607  23.4500   NaN        S  Sharad  
889      0            111369  30.0000  C148        C  Sharad  
890      0            370376   7.7500   NaN        Q  Sharad  

[891 rows x 13 columns]

In [19]:
df['Name']

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

In [20]:
df['Name'][0:15:2] # only 15 record 

0                               Braund, Mr. Owen Harris
2                                Heikkinen, Miss. Laina
4                              Allen, Mr. William Henry
6                               McCarthy, Mr. Timothy J
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                      Sandstrom, Miss. Marguerite Rut
12                       Saundercock, Mr. William Henry
14                 Vestrom, Miss. Hulda Amanda Adolfina
Name: Name, dtype: object

In [21]:
df[['Name']][0:15:2] 

Name
0                             Braund, Mr. Owen Harris
2                              Heikkinen, Miss. Laina
4                            Allen, Mr. William Henry
6                             McCarthy, Mr. Timothy J
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                    Sandstrom, Miss. Marguerite Rut
12                     Saundercock, Mr. William Henry
14               Vestrom, Miss. Hulda Amanda Adolfina

In [22]:
df['Name'][43] 

'Laroche, Miss. Simonne Marie Anne Andree'

In [23]:
df[df['Age'].isnull() == True].index

Int64Index([  5,  17,  19,  26,  28,  29,  31,  32,  36,  42,
            ...
            832, 837, 839, 846, 849, 859, 863, 868, 878, 888],
           dtype='int64', length=177)

In [24]:
len(df[df['Age'].isnull() == True].index)

177

In [25]:
ind = df[df['Age'].isnull() == True].index
df.loc[ind]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked ineuron  
5      male  NaN      0      0      330877   8.4583   NaN        Q  Sharad  
17     male  NaN      0      0      244373  13.0000   NaN        S  Sharad  
19   female  NaN      0      0        2649   7.2250   NaN        C  Sharad  
26     male  NaN      0      0        2631   7.2250   NaN        C  Sharad  
28   female  NaN      0      0      330959   7.8792   NaN        Q  Sharad  
..      ...  ...    ...    ...         ...      ...   ...      ...     ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  Sharad  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  Sharad  
868    male  NaN      0      0      345777   9.5000   NaN        S  Sharad  
878    male  NaN      0      0      349217   7.8958   NaN        S  Sharad  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  Sharad  

[177 rows x 13 columns]

In [26]:
df[df['Age'].isnull()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked ineuron  
5      male  NaN      0      0      330877   8.4583   NaN        Q  Sharad  
17     male  NaN      0      0      244373  13.0000   NaN        S  Sharad  
19   female  NaN      0      0        2649   7.2250   NaN        C  Sharad  
26     male  NaN      0      0        2631   7.2250   NaN        C  Sharad  
28   female  NaN      0      0      330959   7.8792   NaN        Q  Sharad  
..      ...  ...    ...    ...         ...      ...   ...      ...     ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  Sharad  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  Sharad  
868    male  NaN      0      0      345777   9.5000   NaN        S  Sharad  
878    male  NaN      0      0      349217   7.8958   NaN        S  Sharad  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  Sharad  

[177 rows x 13 columns]

In [27]:
df['Age'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

# Que. - Person name who has pay highest fare?

In [28]:
max(df['Fare'])

512.3292

In [29]:
bool_val = df['Fare'] == max(df['Fare'])

In [30]:
df[bool_val]

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   

    ineuron  
258  Sharad  
679  Sharad  
737  Sharad

In [31]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked ineuron  
0      0         A/5 21171   7.2500   NaN        S  Sharad  
1      0          PC 17599  71.2833   C85        C  Sharad  
2      0  STON/O2. 3101282   7.9250   NaN        S  Sharad  
3      0            113803  53.1000  C123        S  Sharad  
4      0            373450   8.0500   NaN        S  Sharad

# Questions

q1.Find out how many male and female passenge was onboarded?

q2. How many survived 

q3. how many casualaties

q4. name the person who is the eldest one

q5. how many passenger do we have in the first, second and third class

q6. how many person we have whose name starts with 's'

q7. try to create a new column which is a summation of 'Sibsp' and 'parch'

q8. how many person do we have below age of 25

q9. how many person died whose age was less than 40

q10. from a cabin column seperate text and numeric value

In [33]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked ineuron  
0      0         A/5 21171   7.2500   NaN        S  Sharad  
1      0          PC 17599  71.2833   C85        C  Sharad  
2      0  STON/O2. 3101282   7.9250   NaN        S  Sharad  
3      0            113803  53.1000  C123        S  Sharad  
4      0            373450   8.0500   NaN        S  Sharad

> q1.Find out how many male and female passenge was onboarded?       
> answer - (Male,Female) = (577, 314)

In [34]:
df_m = df[df['Sex']=='male']
df_F = df[df['Sex']=='female']
len(df_m), len(df_F)

(577, 314)

In [36]:
# solution1 from sudhansu 
df['Sex'] # it is a series 
# group by and count function 
df['Sex'].value_counts() # this method is the combination of count and group  by method

male      577
female    314
Name: Sex, dtype: int64

In [40]:
# solution2  from sudhansu  
df.groupby('Sex')
df.groupby('Sex').count()
df.groupby('Sex')['Sex'].count()

Sex
female    314
male      577
Name: Sex, dtype: int64

> q2. How many survived          
> answer - 342

In [70]:
df_sur = df[df['Survived'] == 1] # this is filter method
len(df_sur)

342

In [42]:
# sudh
df.groupby('Survived')['Survived'].count()

Survived
0    549
1    342
Name: Survived, dtype: int64

In [43]:
df['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

> q3. how many casualaties           
> answer - 549

In [44]:
df_C = df[df['Survived'] == 0]
len(df_C)

549

> q4. name the person who is the eldest one            
> Answer -  Age: 80.0          
Name:Barkworth, Mr. Algernon Henry Wilson

In [48]:
con_80 = df['Age'] == max(df['Age'])
df[con_80]

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   

      Sex   Age  SibSp  Parch Ticket  Fare Cabin Embarked ineuron  
630  male  80.0      0      0  27042  30.0   A23        S  Sharad

In [51]:
# anothe solutjion 
cond = df['Age'] == df['Age'].max()
df[cond]['Name']

630    Barkworth, Mr. Algernon Henry Wilson
Name: Name, dtype: object

> q5. how many passenger do we have in the first, second and third class               
> answer - 

In [52]:
# first class
len(df[df['Pclass']== 1])

216

In [53]:
# second class
len(df[df['Pclass']== 2])

184

In [58]:
# third class
len(df[df['Pclass']== 3])

491

In [60]:
# anothe solut
df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [62]:
df.groupby('Pclass')['Pclass'].count()

Pclass
1    216
2    184
3    491
Name: Pclass, dtype: int64

> q6. how many person we have whose name starts with 's'       
> Answer - 86

In [63]:
series1 = df['Name'].str.startswith('S')
df[series1]
len(df[series1])

86

In [64]:
series1 = df['Name'].str.startswith('s')
df[series1]
len(df[series1])

0

> q7. try to create a new column which is a summation of 'Sibsp' and 'parch'                      
> answer - 

In [65]:
df['summation'] = df.SibSp + df.Parch

In [67]:
df[['SibSp','Parch','summation']]

SibSp  Parch  summation
0        1      0          1
1        1      0          1
2        0      0          0
3        1      0          1
4        0      0          0
..     ...    ...        ...
886      0      0          0
887      0      0          0
888      1      2          3
889      0      0          0
890      0      0          0

[891 rows x 3 columns]

In [70]:
# another solution
df['summation'] = df['SibSp'] + df['Parch']

> q8. how many person do we have below age of 25           
> Answer - 278

In [73]:
condition = df['Age'] < 25 # called filter operation
df1 = df[condition]
len(df1)

278

> q9. how many person died whose age was less than 40                
> Answer -  322

In [75]:
condition = (df['Survived'] == 0)
df_not_s = df[condition]
df_not_s

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
6              7         0       1                   McCarthy, Mr. Timothy J   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch           Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0        A/5 21171   7.2500   NaN        S   
4      male  35.0      0      0           373450   8.0500   NaN        S   
5      male   NaN      0      0           330877   8.4583   NaN        Q   
6      male  54.0      0      0            17463  51.8625   E46        S   
7      male   2.0      3      1           349909  21.0750   NaN        S   
..      ...   ...    ...    ...              ...      ...   ...      ...   
884    male  25.0      0      0  SOTON/OQ 392076   7.0500   NaN        S   
885  female  39.0      0      5           382652  29.1250   NaN        Q   
886    male  27.0      0      0           211536  13.0000   NaN        S   
888  female   NaN      1      2       W./C. 6607  23.4500   NaN        S   
890    male  32.0      0      0           370376   7.7500   NaN        Q   

    ineuron  summation  
0    Sharad          1  
4    Sharad          0  
5    Sharad          0  
6    Sharad          0  
7    Sharad          4  
..      ...        ...  
884  Sharad          0  
885  Sharad          5  
886  Sharad          0  
888  Sharad          3  
890  Sharad          0  

[549 rows x 14 columns]

In [76]:
condition = (df_not_s['Age'] < 40)
df_not_s[condition]

PassengerId  Survived  Pclass                                  Name  \
0              1         0       3               Braund, Mr. Owen Harris   
4              5         0       3              Allen, Mr. William Henry   
7              8         0       3        Palsson, Master. Gosta Leonard   
12            13         0       3        Saundercock, Mr. William Henry   
13            14         0       3           Andersson, Mr. Anders Johan   
..           ...       ...     ...                                   ...   
883          884         0       2         Banfield, Mr. Frederick James   
884          885         0       3                Sutehall, Mr. Henry Jr   
885          886         0       3  Rice, Mrs. William (Margaret Norton)   
886          887         0       2                 Montvila, Rev. Juozas   
890          891         0       3                   Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket    Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.250   NaN        S   
4      male  35.0      0      0            373450   8.050   NaN        S   
7      male   2.0      3      1            349909  21.075   NaN        S   
12     male  20.0      0      0         A/5. 2151   8.050   NaN        S   
13     male  39.0      1      5            347082  31.275   NaN        S   
..      ...   ...    ...    ...               ...     ...   ...      ...   
883    male  28.0      0      0  C.A./SOTON 34068  10.500   NaN        S   
884    male  25.0      0      0   SOTON/OQ 392076   7.050   NaN        S   
885  female  39.0      0      5            382652  29.125   NaN        Q   
886    male  27.0      0      0            211536  13.000   NaN        S   
890    male  32.0      0      0            370376   7.750   NaN        Q   

    ineuron  summation  
0    Sharad          1  
4    Sharad          0  
7    Sharad          4  
12   Sharad          0  
13   Sharad          6  
..      ...        ...  
883  Sharad          0  
884  Sharad          0  
885  Sharad          5  
886  Sharad          0  
890  Sharad          0  

[322 rows x 14 columns]

In [77]:
len(df_not_s[condition])

322

In [79]:
# another solution 
filtr = (df['Survived'] == 0) & (df['Age']<40)
len(df[filtr])

322

> q10. from a cabin column seperate text and numeric value                  
> Answer -

In [85]:
import re 
pattern = re.compile('([a-zA-Z]+)')
df['Cabinstring'] = df['Cabin'].str.extract(pattern)
df['CabinNumeric'] = df['Cabin'].str.replace(pattern,'')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked ineuron  summation  \
0      0         A/5 21171   7.2500   NaN        S  Sharad          1   
1      0          PC 17599  71.2833   C85        C  Sharad          1   
2      0  STON/O2. 3101282   7.9250   NaN        S  Sharad          0   
3      0            113803  53.1000  C123        S  Sharad          1   
4      0            373450   8.0500   NaN        S  Sharad          0   

  Cabinstring CabinNumeric  
0         NaN          NaN  
1           C           85  
2         NaN          NaN  
3           C          123  
4         NaN          NaN

[https://github.com/iNeuronai/same-resume-year-wise](https://github.com/iNeuronai/same-resume-year-wise)  

45;52[https://learn.ineuron.ai/lesson/Pandas-Data-Analysis-Part-2/62eaa6ba766d655b07316a31/course/Full-Stack-Data-Science-BootCamp-2.0/62eaa6ba766d6539c53164bd](https://learn.ineuron.ai/lesson/Pandas-Data-Analysis-Part-2/62eaa6ba766d655b07316a31/course/Full-Stack-Data-Science-BootCamp-2.0/62eaa6ba766d6539c53164bd)

Read all the resume from ineuron git in pdf, word and try to create a dataframe with resume name as index value, and in  column  and i am expection email_id, linked id and skills ?

# 26 Nov 2022

Data source        
         
[https://archive.ics.uci.edu/ml/datasets/bank+marketing](https://archive.ics.uci.edu/ml/datasets/bank+marketing)

In [88]:
import pandas as pd
df = pd.read_csv('bank.csv',sep =';')
df.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [89]:
df.describe()

age       balance          day     duration     campaign  \
count  4521.000000   4521.000000  4521.000000  4521.000000  4521.000000   
mean     41.170095   1422.657819    15.915284   263.961292     2.793630   
std      10.576211   3009.638142     8.247667   259.856633     3.109807   
min      19.000000  -3313.000000     1.000000     4.000000     1.000000   
25%      33.000000     69.000000     9.000000   104.000000     1.000000   
50%      39.000000    444.000000    16.000000   185.000000     2.000000   
75%      49.000000   1480.000000    21.000000   329.000000     3.000000   
max      87.000000  71188.000000    31.000000  3025.000000    50.000000   

             pdays     previous  
count  4521.000000  4521.000000  
mean     39.766645     0.542579  
std     100.121124     1.693562  
min      -1.000000     0.000000  
25%      -1.000000     0.000000  
50%      -1.000000     0.000000  
75%      -1.000000     0.000000  
max     871.000000    25.000000

### Questions about dataset

q1. How many compaign available in this dataset?

q2. How many users do we have with housing and personal loan?

q3. How many do we have whose age is 60 +?

q4. In which month we have targeted most of the customers?

q5.  Which mode of call giving more result?

q6. How many enterprenuer do we have in this dataset?

q7. How many customers do we have with negativa balance?

q8. Prepare a group of data based on education level?

**q1. How many compaign available in this dataset?     
Answer -** There are 

In [93]:
df['campaign'].unique()
len(df['campaign'].unique())

32

In [95]:
# another solution 
df['campaign']
list(df['campaign'])
set(list(df['campaign']))
len(set(list(df['campaign'])))

32

**q2. How many users do we have with housing and personal loan?         
Answer-**       
PL = 691           
HL = 2559        

In [96]:
df_pl = df[df['loan'] == 'yes']
df_pl
len(df_pl)

691

In [97]:
df_hl = df[df['housing'] == 'yes']
df_hl
len(df_hl)

2559

In [98]:
# another solution
df['housing'].value_counts() # answer is yes one 

yes    2559
no     1962
Name: housing, dtype: int64

In [99]:
df['loan'].value_counts() # answer is yes one

no     3830
yes     691
Name: loan, dtype: int64

In [103]:
# having both personal and housing loan
filter1 = (df['housing']=='yes') & (df['loan']=='yes') 
df[filter1]
len(df[filter1])

406

**q3. How many do we have whose age is 60 +?            
Answer -** 127

In [102]:
df_60 = df[df['age']>60]
df_60
len(df_60)

127

**q4. In which month we have targeted most of the customers?              
Answer-**  
There are 521 people subscribed. and mostly in may             
top	may        
freq	93

In [104]:
df_y = df[df['y']== 'yes']  # dataframe where result outcome is yes
df_m = df_y['month']  # select all the month column from df_y
df_m.describe()       # after descibing we get the result

count     521
unique     12
top       may
freq       93
Name: month, dtype: object

In [106]:
# another solution 
df['month'].value_counts()
df['month'].value_counts().index[0]

'may'

**q5. Which mode of call giving more result?                   
Answer**       
top       cellular         
freq           416    

In [108]:
df_y = df[df['y']== 'yes']  # dataframe where result outcome is yes
df_contact = df_y['contact']
df_contact.describe()

count          521
unique           3
top       cellular
freq           416
Name: contact, dtype: object

In [119]:
# another solution by sudh
df_cont_y = df.groupby(['contact','y']).count()
df_cont_y.loc['cellular']
df_cont_y.loc['cellular','yes']
df_cont_y.loc['cellular','yes'][0]

416

**q6. How many enterprenuer do we have in this dataset?           
answer** 168

In [120]:
df_enterprenuer = df[df['job'] == 'entrepreneur']
df_enterprenuer
len(df_enterprenuer)

168

**q7. How many customers do we have with negativa balance?             
Answer**      
366

In [121]:
df_negative_balance = df[df['balance']< 0]
df_negative_balance
len(df_negative_balance)

366

**q8. Prepare a group of data based on education level?         
Answer-**

In [123]:
df['education'].unique()

array(['primary', 'secondary', 'tertiary', 'unknown'], dtype=object)

In [130]:
l = []
for i in df['education'].unique():
    l.append(df[df['education'] == i])

In [131]:
len(l)

4

In [133]:
l[0]

age            job   marital education default  balance housing loan  \
0      30     unemployed   married   primary      no     1787      no   no   
9      43       services   married   primary      no      -88     yes  yes   
18     25    blue-collar    single   primary      no     -221     yes   no   
26     55    blue-collar   married   primary      no      627     yes   no   
36     78        retired  divorced   primary      no      229      no   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4480   23    blue-collar   married   primary      no     1158     yes   no   
4485   53    blue-collar   married   primary      no      238     yes   no   
4486   37    blue-collar   married   primary      no      378     yes   no   
4499   45    blue-collar  divorced   primary      no      942      no   no   
4503   60  self-employed   married   primary      no      362      no  yes   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
0      cellular   19   oct        79         1     -1         0  unknown   no  
9      cellular   17   apr       313         1    147         2  failure   no  
18      unknown   23   may       250         1     -1         0  unknown   no  
26      unknown    5   may       247         1     -1         0  unknown   no  
36    telephone   22   oct        97         1     -1         0  unknown  yes  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4480   cellular   16   apr       743         1     -1         0  unknown   no  
4485   cellular   11   may       238         2    361         5  failure   no  
4486    unknown    9   may       514         1     -1         0  unknown   no  
4499   cellular   21   nov       362         1     -1         0  unknown   no  
4503   cellular   29   jul       816         6     -1         0  unknown  yes  

[678 rows x 17 columns]

In [134]:
l[1]

age          job  marital  education default  balance housing loan  \
1      33     services  married  secondary      no     4789     yes  yes   
4      59  blue-collar  married  secondary      no        0     yes   no   
7      39   technician  married  secondary      no      147     yes   no   
10     39     services  married  secondary      no     9374     yes   no   
11     43       admin.  married  secondary      no      264     yes   no   
...   ...          ...      ...        ...     ...      ...     ...  ...   
4514   38  blue-collar  married  secondary      no     1205     yes   no   
4515   32     services   single  secondary      no      473     yes   no   
4516   33     services  married  secondary      no     -333     yes   no   
4518   57   technician  married  secondary      no      295      no   no   
4519   28  blue-collar  married  secondary      no     1137      no   no   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
1     cellular   11   may       220         1    339         4  failure  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
7     cellular    6   may       151         2     -1         0  unknown  no  
10     unknown   20   may       273         1     -1         0  unknown  no  
11    cellular   17   apr       113         2     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4514  cellular   20   apr        45         4    153         1  failure  no  
4515  cellular    7   jul       624         5     -1         0  unknown  no  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  

[2306 rows x 17 columns]

In [137]:
l[2]

age            job   marital education default  balance housing loan  \
2      35     management    single  tertiary      no     1350     yes   no   
3      30     management   married  tertiary      no     1476     yes  yes   
5      35     management    single  tertiary      no      747      no   no   
6      36  self-employed   married  tertiary      no      307     yes   no   
8      41   entrepreneur   married  tertiary      no      221     yes   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4501   34     management   married  tertiary      no      297     yes   no   
4506   42     unemployed  divorced  tertiary      no     -166      no   no   
4509   51     technician   married  tertiary      no     2506      no   no   
4517   57  self-employed   married  tertiary     yes    -3313     yes  yes   
4520   44   entrepreneur    single  tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
5     cellular   23   feb       141         2    176         3  failure  no  
6     cellular   14   may       341         1    330         2    other  no  
8      unknown   14   may        57         2     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4501  cellular   26   aug        63         4     -1         0  unknown  no  
4506  cellular   29   aug        85         4     -1         0  unknown  no  
4509  cellular   30   nov       210         3     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[1350 rows x 17 columns]

In [136]:
l[3]

age          job   marital education default  balance housing loan  \
20     38   management  divorced   unknown      no        0     yes   no   
27     67      retired   married   unknown      no      696      no   no   
49     61       admin.   married   unknown      no     4629     yes   no   
132    43  blue-collar   married   unknown     yes     -715     yes   no   
133    48       admin.   married   unknown      no        0     yes   no   
...   ...          ...       ...       ...     ...      ...     ...  ...   
4337   73      retired   married   unknown      no      519      no   no   
4342   50  blue-collar   married   unknown      no     2284     yes   no   
4428   48  blue-collar   married   unknown      no      439     yes   no   
4471   59   management   married   unknown      no     3534      no   no   
4508   42       admin.   married   unknown      no      642     yes  yes   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
20     cellular   18   nov        96         2     -1         0  unknown   no  
27    telephone   17   aug       119         1    105         2  failure   no  
49     cellular   27   jan       181         1     92         1  success  yes  
132     unknown   30   may        67         3     -1         0  unknown   no  
133    cellular    8   may        85         1    168         2  failure   no  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4337  telephone   16   oct       434         1     57         1  failure  yes  
4342  telephone   31   jul      1088        17     -1         0  unknown  yes  
4428   cellular    2   feb       212         1    249         2    other   no  
4471   cellular   21   nov       216         4     -1         0  unknown   no  
4508    unknown   16   may       509         2     -1         0  unknown   no  

[187 rows x 17 columns]